cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/mds/AbstractMDSUpgrader.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.datafabric.dataset.service.mds;
+
+import co.cask.cdap.common.io.Locations;
+import com.google.common.collect.ImmutableList;
+import org.apache.twill.filesystem.Location;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * An Abstract class which provides common methods and classes need by Dataset MDS to be upgraded
+ */
+public abstract class AbstractMDSUpgrader {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractMDSUpgrader.class);
+
+
+  /**
+   * An abstract method which should be implemented by dataset MDS which needs to be upgraded
+   * @throws Exception
+   */
+  public abstract void upgrade() throws Exception;
+
+  /**
+   * Renames the old location to new location if old location exists and the new one does not
+   *
+   * @param oldLocation the old {@link Location}
+   * @param newLocation the new {@link Location}
+   * @return new location if and only if the file or directory is successfully moved; null otherwise.
+   * @throws IOException
+   */
+  @Nullable
+  protected Location renameLocation(Location oldLocation, Location newLocation) throws IOException {
+    // if the newLocation does not exists or the oldLocation does we try to rename. If either one of them is false then
+    // the underlying call to renameTo will throw IOException which we propagate.
+    if (!newLocation.exists() || oldLocation.exists()) {
```

Comment:
My intention was: 
- if newLocation does no exists first case evaluates to true and we get in if loop. Now if oldLocation exists all good move the file. If the oldLocation does not exists renameTo will throw exception which we rethrow. Where did the files go ? :)

- If newLocation exists second case gets evaluated. OldLocation does not exists all good files were already moved :) If old does exists the renameTo calls with overwrite Options.Rename.OVERWRITE which is true I thought that might be a good idea but today we did noticed new logs coming up in old dirs so doesn't look like its a good idea. 

I guess && does makes more sense since these paths are generally from specs  so we do know that they will exists